lib/oidcstrategy.js

```diff
@@ -752,4 +574,468 @@ Strategy.prototype._shouldLoadUserProfile = function shouldLoadUserProfile(issue
   return done(null, !skip);
 };
 
+/**
+ * validate id_token, and pass the validated claims and the payload to callback
+ * if code (resp. access_token) is provided, we will validate the c_hash (resp at_hash) as well
+ *
+ * @param {String} id_token
+ * @param {Object} req
+ * @param {Function} callback
+ * @param {String} code (if you want to validate c_hash)
+ * @param {String} access_token (if you want o validate at_hash)
+ */
+Strategy.prototype._validateIdToken = function validateIdToken(id_token, code, access_token, req, callback) {
+  const self = this;
+
+  // decode id_token
+  const decoded = jws.decode(id_token);
+  if (decoded == null)
+    return self. fail(null, false, 'Invalid JWT token');
+  log.info('token decoded: ', decoded);
+
+  // get Pem Key
+  var PEMkey = undefined;
+  if (decoded.header.x5t) {
+    PEMkey = self.metadata.generateOidcPEM(decoded.header.x5t);
+  } else if (decoded.header.kid) {
+    PEMkey = self.metadata.generateOidcPEM(decoded.header.kid);
+  } else {
+    return self.fail('We did not reveive a token we know how to validate');
+  }
+
+  var options = self._options;
+
+  // since the id_token is for itself, so the id_token is the clientID by default
+  options.audience = options.clientID;
+
+  // verify id_token signature and claims
+  return jwt.verify(id_token, PEMkey, options, (err, jwtClaims) => {
+    if (err)
+      return self.fail("cannot verify id token");
+    log.info("Claims received: ", jwtClaims);
+
+    // jwt checks the 'nbf', 'exp', 'aud', 'iss' claims
+    // there are a few other things we will check below
+
+    // we don't allow multiple audiences in id_token
+    if (Array.isArray(jwtClaims.aud) && jwtClaims.aud.length > 1)
+      return self.fail('we do not allow multiple audiences in id_token');
+
+    // check the nonce in claims
+    var nonceCheckResult = nonceHandler.verifyNonce(req, self._key, jwtClaims.nonce);
+    if (!nonceCheckResult.valid)
+      return self.fail(nonceCheckResult.errorMessage);
+
+    // check c_hash
+    if (code && jwtClaims.c_hash && !utils.checkHashValueRS256(code, jwtClaims.c_hash))
+      return self.fail("invalid c_hash");
+
+    // check at_hash
+    if (access_token && jwtClaims.at_hash && !utils.checkHashValueRS256(access_token, jwtClaims.at_hash))
+      return self.fail("invalid at_hash");
+
+    // return jwt claims and jwt claims string
+    var idTokenSegments = id_token.split('.');
```

Comment:
line 74 of index.js in jsonwebtoken:

```
  var hasSignature = parts[2].trim() !== '';

  if (!hasSignature && secretOrPublicKey){
    return done(new JsonWebTokenError('jwt signature is required'));
  }
```
the error raised will go into the 'err' in the callback in our code:
```
return jwt.verify(id_token, PEMkey, options, (err, jwtClaims) => {
```